### PR TITLE
Wire together LRU garbage collection

### PR DESCRIPTION
**Pull request by gsoltis**
Port of https://github.com/firebase/firebase-android-sdk/pull/68

Enables LRU garbage collection, sets the default theshold at 40mb.

### CODE REVIEW

File: packages/firebase/index.d.ts [hunk 1]
```diff
@@ -768,6 +770,17 @@ declare namespace firebase.firestore {
...
+    /**
+     * An approximate cache size threshold for the on-disk data. If the cache grows beyond this
+     * size, Firestore will start removing data that hasn't been recently used. The size is not a
+     * guarantee that the cache will stay below that size, only that if the cache exceeds the given
+     * size, cleanup will be attempted.
+     *
+     * The default value is 40 MB. The threshold must be set to at least 1 MB, and can be set to
+     * CACHE_SIZE_UNLIMITED to disable garbage collection.
+     */
+    cacheSizeBytes?: number;
```
Comment (Feiyang1) [inline packages/firebase/index.d.ts]: Can you please also update the externs file?

Comment (gsoltis) [reply packages/firebase/index.d.ts]: Done.

File: packages/firestore/src/api/database.ts [hunk 2]
```diff
@@ -418,7 +454,10 @@ follow these steps, YOUR APP MAY BREAK.`);
       this._config.credentials,
       this._queue
     );
-    return this._firestoreClient.start(persistenceSettings);
+    return this._firestoreClient.start(
+      persistenceSettings,
+      this._config.settings.cacheSizeBytes
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply packages/firestore/src/api/database.ts]: PR here: https://github.com/firebase/firebase-js-sdk/pull/1411

File: packages/firestore/test/unit/local/lru_garbage_collector.test.ts [hunk 3]
```diff
@@ -983,4 +986,50 @@ function genericLruGarbageCollectorTests(
     );
     expect(finalCacheSize).to.be.lessThan(cacheSize);
   });
+
+  it('caps sequence numbers to collect', async () => {
+    //await persistence.shutdown();
```
Comment (mikelehen) [inline packages/firestore/test/unit/local/lru_garbage_collector.test.ts]: What's this?

File: packages/firestore/src/local/lru_garbage_collector.ts [hunk 4]
```diff
@@ -146,21 +147,25 @@ class RollingSequenceNumberBuffer {
...
+ * `false` if collection was skipped (either it is disabled or the cache size
+ * has not hit the threshold). If collection ran, the other fields will be
+ * filled in with the details of the results.
+ */
 export type LruResults = {
-  hasRun: boolean;
-  sequenceNumbersCollected: number;
-  targetsRemoved: number;
-  documentsRemoved: number;
+  readonly hasRun: boolean;
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply packages/firestore/src/local/lru_garbage_collector.ts]: Renamed to `didRun`.

File: packages/firestore/test/unit/local/lru_garbage_collector.test.ts (hunk 3 above)
Comment (gsoltis) [reply packages/firestore/test/unit/local/lru_garbage_collector.test.ts]: Removed.

Maybe irrelevant details, but it turns out that with multi-tab, shutting down the persistence layer outside of the async queue can lead to race conditions. The test teardown and initialize methods now do the right thing.

File: packages/firestore/src/api/database.ts [hunk 5]
```diff
@@ -101,6 +102,15 @@ import {
 const DEFAULT_HOST = 'firestore.googleapis.com';
 const DEFAULT_SSL = true;
 const DEFAULT_TIMESTAMPS_IN_SNAPSHOTS = false;
+const MINIMUM_CACHE_BYTES = 1 * 1024 * 1024;
+const DEFAULT_CACHE_SIZE_BYTES = LruParams.DEFAULT_CACHE_SIZE_BYTES;
```
[the remark this answers is not in the document]
Comment (gsoltis) [reply packages/firestore/src/api/database.ts]: Done. I removed the redeclaration of the constants and instead just used the ones from `LruParams` directly.

**Pull request merged**